cbmc/proofs/IotMqtt_Disconnect/IotMqtt_Disconnect_harness.c

```diff
@@ -0,0 +1,133 @@
+ * list are accessed after the remove: We free all eleents on the
+ * original list, so that any later access will be caught as a
+ * use-after-free error.
+ */
+
+typedef bool ( *MatchFunction_t )( const IotLink_t * const pOperationLink,
+                                   void * pCompare );
+typedef void ( *FreeElementFunction_t )( void * pData );
+
+void IotListDouble_RemoveAllMatches( const IotListDouble_t * const pList,
+                                     MatchFunction_t isMatch,
+                                     void * pMatch,
+                                     FreeElementFunction_t freeElement,
+                                     size_t linkOffset )
+{
+    free_IotMqttSubscriptionList( pList );
+    allocate_IotMqttSubscriptionList( pList, SUBSCRIPTION_COUNT_MAX - 1 );
+}
+
+void IotListDouble_RemoveAll( const IotListDouble_t * const pList,
+                              FreeElementFunction_t freeElement,
+                              size_t linkOffset )
+{
+    free_IotMqttSubscriptionList( pList );
+    allocate_IotMqttSubscriptionList( pList, SUBSCRIPTION_COUNT_MAX - 1 );
+}
+
+/****************************************************************/
+
+/**
+ * We assume the semaphore operations are memory safe.
+ *
+ * We abstract the semaphores because we are doing sequential proof.
+ * But the semaphore API assures us that TimedWait called after Post will
+ * never fail. Our abstraction of the semaphores models this behavior.
+ *
+ * Our abstraction is safe because the Disconnect method invokes the
+ * semaphore methods on the semaphore associated with the Disconnect
+ * packet.  The Disconnect method is the only code that accesses this
+ * semaphore.  This justifies this simple semaphore model of checking
+ * for Post before TimedWait.
+ */
+
+static unsigned int flagSemaphore;
+
+void IotSemaphore_Post( IotSemaphore_t * pSemaphore )
+{
+    assert( pSemaphore != NULL );
+    flagSemaphore++;
+}
+
+bool IotSemaphore_TimedWait( IotSemaphore_t * pSemaphore,
+                             uint32_t timeoutMs )
+{
+    assert( pSemaphore != NULL );
+
+    if( flagSemaphore > 0 )
+    {
+        flagSemaphore--;
+        return true;
+    }
+
+    return false;
+}
+
+/****************************************************************/
+
+void harness()
+{
+    /* initialize semaphore flag */
+    flagSemaphore = 0;
+
+    IotMqttConnection_t mqttConnection = allocate_IotMqttConnection( NULL );
+    uint32_t flags;
+
+    __CPROVER_assume( mqttConnection != NULL );
+    ensure_IotMqttConnection_has_lists( mqttConnection );
+    __CPROVER_assume( valid_IotMqttConnection( mqttConnection ) );
+    __CPROVER_assume( mqttConnection->references > 0 );
+    __CPROVER_assume( IMPLIES( mqttConnection->pingreq.u.operation.periodic.ping.keepAliveMs != 0,
+                               mqttConnection->references > 1 ) );
+    __CPROVER_assume( mqttConnection->pNetworkInterface != NULL );
```

Comment:
Add a comment that `IotMqtt_Disconnect` must be resilient enough to ensure this assumptions.

Reply:
I've labeled this as a list of proof assumptions.  To say more, well, what makes this proof special?  This statement could be true for all of our proofs.